text/0000-try-trait.md

```diff
@@ -0,0 +1,553 @@
+`Carrier` was implemented separately for each type (e.g., `Result` and
+`Option`). It allowed any kind of carrier to be converted to any other
+kind of carrier, which fails to preserve the "semantic equivalent"
+property.  It would also interact poorly with type inference, as
+discussed in the "Detailed Design" section.
+
+### Traits implemented over higher-kinded types
+
+The "semantic equivalent" property might suggest that the `Carrier`
+trait ought to be defined over higher-kinded types (or generic
+associated types) in some form. The most obvious downside of such a
+design is that Rust does not offer higher-kinded types nor anything
+equivalent to them today, and hence we would have to block on that
+design effort. But it also turns out that HKT is
+[not a particularly good fit for the problem](https://github.com/rust-lang/rust/pull/35056#issuecomment-240129923). To
+start, consider what "kind" the `Self` parameter on the `Try` trait
+would have to have.  If we were to implement `Try` on `Option`, it
+would presumably then have kind `type -> type`, but we also wish to
+implement `Try` on `Result`, which has kind `type -> type ->
+type`. There has even been talk of implementing `Try` for simple types
+like `bool`, which simply have kind `type`. More generally, the
+problems encountered are quite similar to the problems that
+[Simon Peyton-Jones describes in attempting to model collections using HKT](https://github.com/rust-lang/rust/pull/35056#issuecomment-240129923):
+we wish the `Try` trait to be implemented in a great number of
+scenarios.  Some of them, like converting `Result<T,E>` to
+`Result<U,F>`, allow for the type of the success value and the error
+value to both be changed, though not arbitrarily (subject to the
+`From` trait, in particular).  Others, like converting `Option<T>` to
+`Option<U>`, allow only the type of the success value to change,
+whereas others (like converting `bool` to `bool`) do not allow either
+type to change.
+
+### What to name the trait
+
+A number of names have been proposed for this trait. The original name
+was `Carrier`, as the implementing type was the "carrier" for an error
+value. A proposed alternative was `QuestionMark`, named after the
+operator `?`. However, the general consensus seemed to be that since
+Rust operator overloading traits tend to be named after the
+*operation* that the operator performed (e.g., `Add` and not `Plus`,
+`Deref` and not `Star` or `Asterix`), it was more appropriate to name
+the trait `Try`, which seems to be the best name for the operation in
+question.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+**We need to resolve the interactions with type inference.** It is
+important that expressions like `vec.iter().map(|e| ...).collect()?`
+are able to infer the type of `collect()` from context. This may
+require small tweaks to the `Try` trait, though the author considers
+that unlikely.
+
```

Comment:
> It would be diligent to preserve the “cute” collect trick when migrating from try to ? but it's not a pressing issue like error conversion.

I'm not sure what you mean. When I write code using `?` today, I still have to use `try!` whenever there is a call to `collect()`. This happens to me a lot. Seems like a problem.

> Nesting ? doesn't work without type annotations on closures or what it may be; that is not good for the intended design of ? and catch.

But this is also true for `try!`, right? It's not clear to me why this is a bigger problem for `?` and `catch` than `try!`. Closures in particular tend to get their return types inferred from context. In any case, I agree it's a problem we should try to address, though the best solution is unclear; it seems to require some sort of inference tricks.

In any case, it's not like we can remove error conversion! It's already stabilized, and we want to be able to serve as a drop-in replacement for `try!`.